### ISSUE TITLE
False positive with -Wformat-nonliteral  when using brace initialization

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#58900]
Hi,

We find that -Wformat-nonliteral does not recognize string literals if they are initialized with brace initialization. Example:

```cpp
#include <cstdio>

int main()
{
    constexpr const char* fmt = "%d";  // OK
    char data[100];
    sprintf(data, fmt, 123);  

    constexpr const char* fmt2{"%d"};  // NOK
    sprintf(data, fmt2, 123);  
}
```

Repro: https://godbolt.org/z/Yn14aqzvb

**Comment by inclyc** [comment]
`InitListExpr` is currently not handled in Sema subsystem. We may need to evaluate these exprs like `{'a', 'b', 'c'}` too invasively to support. Is there any case that we have to handle this in the real world? I'm interested to know if there is a real practical scenario where such brace expressions are necessary instead of using string literals directly?

Note: There is a patch might fix this issue but abandoned: https://reviews.llvm.org/D131314.

May duplicate to: https://github.com/llvm/llvm-project/issues/56925

**Comment by inclyc** [comment]
CC @AaronBallman

**Comment by carlosgalvezp** [comment]
Should this be a `InitListExpr` though? I suppose it's a general Sema question as I've seen it in any kind of brace initialization.

In the real world we don't use `{'a', 'b', 'c'}`, but some guidelines like AUTOSAR mandate the use of brace initialization instead of = initialization, to prevent narrowing.

>Rule A8-5-2 (required, implementation, automated) 
>Braced-initialization {}, without equals sign, shall be used for variable initialization.

Therefore our code needs to look like:
```cpp
char const* x{"foo"};
```
instead of:
```cpp
char const* x = "foo";
```

We can of course disable the Clang warning but it's quite useful and seems like it should support both cases.

**Comment by inclyc** [comment]
> Should this be a `InitListExpr` though? I suppose it's a general Sema question as I've seen it in any kind of brace initialization.
> 
> In the real world we don't use `{'a', 'b', 'c'}`, but some guidelines like AUTOSAR mandate the use of brace initialization instead of = initialization, to prevent narrowing.
> 
> > Rule A8-5-2 (required, implementation, automated)
> > Braced-initialization {}, without equals sign, shall be used for variable initialization.
> 
> Therefore our code needs to look like:
> 
> ```c++
> char const* x{"foo"};
> ```
> 
> instead of:
> 
> ```c++
> char const* x = "foo";
> ```
> 
> We can of course disable the Clang warning but it's quite useful and seems like it should support both cases.

Thx for your early reply! In this scenario I think it is reasonable to deal with `{"foobar"}` though. And this issue is not duplicated.

**Comment by inclyc** [comment]
> Should this be a InitListExpr though? I suppose it's a general Sema question as I've seen it in any kind of brace initialization.

It is, you can inspect the ast-dump [here](https://godbolt.org/z/noEcddhEr).

**Comment by inclyc** [comment]
Patch https://reviews.llvm.org/D137839

**Comment by carlosgalvezp** [comment]
Amazing, thanks for the quick fix!

> It is, you can inspect the ast-dump [here](https://godbolt.org/z/noEcddhEr).

Yes, I meant that it feels strange that it's classified as such, since there is no list here - it's just brace initialization (I think the correct term is "uniform" initialization). But you guys probably know much better :)
